### PR TITLE
Add Writer interface to blobstore

### PR DESCRIPTION
**Pull request by tempoz**
Add interface to blobstore to get an `io.Writer` so we can stream to blobstore.

---

**Version bump**: Minor

### CODE REVIEW

File: server/backends/blobstore/util/util.go
Comment (bduffany) [inline server/backends/blobstore/util/util.go:144]: we might want to omit these Write metrics for now -- the metrics could get skewed in the case where the gzipWriter is doing internal buffering/async flushing (and the metrics would show that the Write() completes in nanoseconds, even the bytes have just been moved from one in-memory buffer to another).

Another option is to do it in `Close()` but that is also not ideal because it depends on client upload latency+throughput (while the existing usages of this metric only depend on the server -> blobstore latency+throughput).

File: server/testutil/mockstore/mockstore.go
Comment (bduffany) [inline server/testutil/mockstore/mockstore.go:100]: nit: if this is updated to be a CommitWriteCloser then this line would probably fit better in Commit()

File: server/backends/blobstore/aws/aws.go
Comment (bduffany) [inline server/backends/blobstore/aws/aws.go:266]: In these async goroutine funcs, we should begin the goroutine with `defer pr.Close()` as the first line, so that if the async blobstore upload fails, any subsequent writes to the returned CompressedBlobstoreWriter would fail (due to the pipe being closed), rather than getting stuck.

Example: https://go.dev/play/p/2TQ4JP8y10w

**Review by bduffany — CHANGES_REQUESTED** [review]

File: server/backends/blobstore/aws/aws.go
Comment (sluongng) [reply server/backends/blobstore/aws/aws.go:269]: I sent out https://github.com/buildbuddy-io/buildbuddy/pull/3736

@tempoz I intend to merge it only after your series lands. Let me know if you prefer to have it merged first.

File: server/testutil/mockstore/mockstore.go
Comment (tempoz) [reply server/testutil/mockstore/mockstore.go:109]: Done!

File: server/backends/blobstore/util/util.go
Comment (bduffany) [reply server/backends/blobstore/util/util.go]: Ah gotcha, I forgot we were already using gzip to store things in blobstore.

File: server/backends/blobstore/util/util.go
Comment (bduffany) [inline server/backends/blobstore/util/util.go:157]: if `c.zw.Close()` fails, should we assume something went wrong and abort - i.e. return the error instead of calling `committer.Commit()` ?

File: server/backends/blobstore/aws/aws.go
Comment (bduffany) [inline server/backends/blobstore/aws/aws.go:284]: Since the `AsyncCloser` is only being used in these `committer` funcs now, and the `Close()` logic is relatively short (*), wdyt about inlining the logic here and removing the AsyncCloser class?

(*) the `Close()` logic could also be a little shorter:

```
	if err := pw.Close(); err != nil {
    // This next if {} block can be removed
    // since errch is buffered
		if err := <-a.ErrorChannel; err != nil {
			return err
		}
		return err
	}
	return <-errch
```

File: server/backends/blobstore/aws/aws.go
Comment (vadimberezniker) [inline server/backends/blobstore/aws/aws.go:283]: This functionality is already available via `ioutil.NewCustomCommitWriteCloser`. 
You could have the compressed writer be responsible only for the compression and then wrap it here with additional commit/close funcs. 
```	
    cwc := ioutil.NewCustomCommitWriteCloser(util.NewCompressedBlobStoreWriter(ctx, pw, awsS3Label))
	cwc.CommitFn = func(i int64) error {
		// ...
	}
	cwc.CloseFn = func() error {
		// ...
	}
```

File: server/backends/blobstore/util/util.go
Comment (vadimberezniker) [inline server/backends/blobstore/util/util.go:164]: This is too fine grained for a span. Normally a span represents a slightly higher level operation (i.e. "write blob") and individual writes might be represented as span events.

File: server/backends/blobstore/util/util.go
Comment (tempoz) [reply server/backends/blobstore/util/util.go]: Good point, done.

File: server/backends/blobstore/aws/aws.go
Comment (vadimberezniker) [inline server/backends/blobstore/aws/aws.go:297]: You can return the error directly instead of storing it. Since errch is buffered, there's no harm if you return before reading from the channel.

File: server/backends/blobstore/aws/aws.go
Comment (vadimberezniker) [inline server/backends/blobstore/aws/aws.go:300]: Same here.

File: server/backends/blobstore/gcs/gcs.go
Comment (vadimberezniker) [inline server/backends/blobstore/gcs/gcs.go:149]: It's not 100% clear from the GCS client docs, but cancelling the context automatically closes the writer so it is likely safer not to try to call `Close` as well since there might be a race between the context cancellation and the close that could lead to the write succeeding.

File: server/backends/blobstore/util/util.go
Comment (vadimberezniker) [reply server/backends/blobstore/util/util.go]: If you do want to bring it back in the future, the way to do it is create a new span in `Writer` and then call `span.AddEvent` for each individual write. This allows you to add fine grained timing information w/o creating new spans.

**Review by vadimberezniker — APPROVED** [review]

File: server/backends/blobstore/aws/aws.go
Comment (bduffany) [inline server/backends/blobstore/aws/aws.go:292]: nit: this channel receive can be deleted since the channel is buffered. it'll just collect the write error once line 280 is hit, then get garbage collected.

**Review by bduffany — APPROVED** [review]

File: server/backends/blobstore/aws/aws.go
Comment (tempoz) [reply server/backends/blobstore/aws/aws.go]: Done.

File: server/backends/blobstore/gcs/gcs.go
Comment (tempoz) [reply server/backends/blobstore/gcs/gcs.go]: Thanks, didn't realize; fixed!

File: server/backends/blobstore/aws/aws.go
Comment (tempoz) [reply server/backends/blobstore/aws/aws.go]: Thanks, didn't know; done.

**Pull request merged**
